### PR TITLE
Spec Improvements

### PR DESCRIPTION
**Pull request by maxnz**
- Increase readability of specs by moving tests to the file under `files:` they are associated with
  - Both listing tests with the file and under a separate `tests` section can be done in the same spec, even for the same file
- `supporting` tag added
  - Same functionality as `inputs`, added to make specs clearer
  - Both `supporting` and `inputs` can be used in the same spec
- Supporting files can be specified with a dict, in addition to a string or list as before:
```yaml
supporting:
  - file1.txt                   # copy file to homework directory
  - [ file2.txt, file2a.txt ]   # copy file to homework directory with rename/relocation
  - file: file3.txt             # same as above but more verbose
    destination: file3a.txt
```

- Spec versions can now be combined, no need to differentiate with `spec_version: 3` anymore.
- Remove an extraneous guarded return that doesn't do anything anymore.

### CODE REVIEW

File: stograde/student/markdownify/process_file.py [hunk 1]
```diff
@@ -139,14 +141,13 @@ def process_file(filename, *, steps, options, spec, cwd, supporting_dir, interac
...
-    should_continue = test_file(filename,
-                                spec=spec,
-                                results=results,
-                                options=options,
-                                cwd=cwd,
-                                supporting_dir=supporting_dir,
-                                interact=interact)
-    if not should_continue:
-        return results
+    test_file(filename,
```
Comment (hawkrives) [inline stograde/student/markdownify/process_file.py]: Why did you remove the return value?

File: README.md [hunk 2]
```diff
@@ -168,10 +168,9 @@ compilers:
   - &cpp 'g++ --std=c++11 $@ -o $@.exec'
 
 files:
-  - [ types.cpp, *cpp ]
-
-tests:
-  - [ types.cpp, $@.exec ]
+  - file: types.cpp
+    commands: *cpp
+    tests: $@.exec
```
Comment (hawkrives) [inline README.md]: The only thing I worry about is that I seem to remember some caees where it would be non-obvious what file a test belonged under.

As long as we support the old syntax, I'm fine with that.

Also, what if you need multiple tests for a given file?

Comment (maxnz) [reply README.md]: The `tests` tag supports having a string or an array of strings just like the `commands` tag.

I was having trouble getting the two syntaxes to combine into one iterable array but I'll put a little more work into it later today

File: stograde/student/markdownify/process_file.py (hunk 1 above)
Comment (maxnz) [reply stograde/student/markdownify/process_file.py]: The return wasn't being used for anything once that check was removed. And the check did the same thing no matter if it evaluated to true or false

File: README.md (hunk 2 above)
Comment (maxnz) [reply README.md]: Do you think we could get the wiki feature of GitHub turned on so we could go more in depth on specific parts, such as all the options for specs, without having an excessively long README?

Comment (hawkrives) [reply README.md]: Idk. I can, but I like having the docs versioned in the repo. 

What about making a `docs/` folder in the repo, with markdown files inside?

Comment (maxnz) [reply README.md]: That sounds like a good idea. I'll start working on that in a separate PR

File: stograde/student/markdownify/process_file.py [hunk 3]
```diff
@@ -55,10 +55,11 @@ def compile_file(filename, *, steps, results, supporting_dir, basedir, spec_id):
...
 
-def test_file(filename, *, spec, results, options, cwd, supporting_dir, interact):
-    tests = flatten([test_spec['commands']
-                     for test_spec in spec.get('tests', {})
-                     if test_spec['filename'] == filename])
+def test_file(filename, *, spec, tests, results, options, cwd, supporting_dir, interact):
+    tests = tests if spec.get('spec_version', 2) >= 3 else \
+        flatten([test_spec['commands']
+                 for test_spec in spec.get('tests', {})
+                 if test_spec['filename'] == filename])
```
Comment (hawkrives) [inline stograde/student/markdownify/process_file.py]: Let's split this into a few lines…

```suggestion
	if spec.get('spec_version', 2) >= 3:
		tests = tests
	else:
    	tests = flatten([test_spec['commands']
                 		 for test_spec in spec.get('tests', {})
                 		 if test_spec['filename'] == filename])
```

File: stograde/student/markdownify/process_file.py [hunk 4]
```diff
@@ -99,9 +100,10 @@ def test_file(filename, *, spec, results, options, cwd, supporting_dir, interact
     return True
 
 
-def process_file(filename, *, steps, options, spec, cwd, supporting_dir, interact, basedir, spec_id,
-                 skip_web_compile):
+def process_file(filename, *, steps, tests, options, spec, cwd, supporting_dir, interact, basedir,
+                 spec_id, skip_web_compile):
     steps = steps if isinstance(steps, Iterable) else [steps]
+    tests = tests if isinstance(tests, Iterable) else [tests]
```
Comment (hawkrives) [inline stograde/student/markdownify/process_file.py]: Unfortunately, strings are iterable.

How about flipping the sides?

```suggestion
    tests = [tests] if isinstance(tests, str) else tests
```

File: stograde/student/markdownify/process_file.py (hunk 3 above)
Comment (maxnz) [reply stograde/student/markdownify/process_file.py]: I figured out a better way that combines the two. I was also able to make it so the `spec_version: 3` line is not necessary anymore, both styles can now be combined in the same spec.

File: stograde/student/markdownify/process_file.py (hunk 4 above)
Comment (maxnz) [reply stograde/student/markdownify/process_file.py]: Should we do this for `steps` too? Though in theory both should already be arrays because of how they are cached, right? Or am I missing something there

**Comment by hawkrives** [comment]
Random drive by comment

I'd like to pick one of destination or dest and only support one

File: stograde/student/markdownify/process_file.py (hunk 4 above)
Comment (maxnz) [reply stograde/student/markdownify/process_file.py]: As far as I can tell, lines [97-99](https://github.com/StoDevX/stograde/blob/aa6115bfb60cc713525f8f9a8b36d162e9ef109c/stograde/specs/cache.py#L97) and [113](https://github.com/StoDevX/stograde/blob/aa6115bfb60cc713525f8f9a8b36d162e9ef109c/stograde/specs/cache.py#L113) in `cache.py` should take care of this, right?

Comment (hawkrives) [reply stograde/student/markdownify/process_file.py]: looks good to me
